xtask/src/main.rs

```diff
@@ -0,0 +1,388 @@
+mod build;
+mod command;
+
+use anyhow::bail;
+use core::fmt;
+use std::{
+    error::Error,
+    ffi::OsString,
+    path::{Path, PathBuf},
+    process,
+    process::ExitStatus,
+    str,
+};
+use structopt::StructOpt;
+
+use crate::{
+    build::{build_hexpath, compare_builds, init_build_dir},
+    command::{run_command, run_successful, BuildMode, CargoCommand},
+};
+
+const ARMV6M: &str = "thumbv6m-none-eabi";
+const ARMV7M: &str = "thumbv7m-none-eabi";
+
+#[derive(Debug, StructOpt)]
+struct Options {
+    #[structopt(short, long)]
+    target: String,
+}
+
+#[derive(Debug)]
+pub struct RunResult {
+    exit_status: ExitStatus,
+    output: String,
+}
+
+#[derive(Debug)]
+enum TestRunError {
+    FileCmpError { file_1: String, file_2: String },
+    PathConversionError(OsString),
+    CommandError(RunResult),
+    IncompatibleCommand,
+}
+
+impl fmt::Display for TestRunError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            TestRunError::FileCmpError { file_1, file_2 } => {
+                write!(f, "Differing output in Files: {} {}", file_1, file_2)
+            }
+            TestRunError::CommandError(e) => {
+                write!(
+                    f,
+                    "Command failed with exit status {}: {}",
+                    e.exit_status, e.output
+                )
+            }
+            TestRunError::PathConversionError(p) => {
+                write!(f, "Can't convert path from `OsString` to `String`: {:?}", p)
+            }
+            TestRunError::IncompatibleCommand => {
+                write!(f, "Can't run that command in this context")
+            }
+        }
+    }
+}
+
+impl Error for TestRunError {}
+
+fn main() -> anyhow::Result<()> {
+    // if there's an `xtask` folder, we're *probably* at the root of this repo (we can't just
+    // check the name of `env::current_dir()` because people might clone it into a different name)
+    let probably_running_from_repo_root = Path::new("./xtask").exists();
```

Comment:
Might consider checking for `.git`?

Reply:
YOLO, missed this comment. We can fix later :)

Reply:
> Might consider checking for `.git`?

I intentionally didn't do that in case someone has some wild submodule-ing going on :)